raiden/utils/signing.py

```diff
@@ -0,0 +1,97 @@
+from coincurve import PrivateKey, PublicKey
+from eth_utils import decode_hex, keccak, remove_0x_prefix, to_bytes
+from web3.utils.abi import map_abi_data
+from web3.utils.encoding import hex_encode_abi_type
+from web3.utils.normalizers import abi_address_to_hex
+
+from raiden.exceptions import InvalidSignature
+from raiden.utils.typing import Address, Callable, Optional, Union
+
+sha3 = keccak
+Hasher = Optional[Callable[[bytes], bytes]]
+
+
+def eth_sign_sha3(data: bytes) -> bytes:
+    """
+    eth_sign/recover compatible hasher
+    Prefixes data with "\x19Ethereum Signed Message:\n<len(data)>"
+    """
+    prefix = b'\x19Ethereum Signed Message:\n'
+    if not data.startswith(prefix):
+        data = prefix + b'%d%s' % (len(data), data)
+    return sha3(data)
+
+
+def public_key_to_address(public_key: Union[PublicKey, bytes]) -> Address:
+    """ Converts a public key to an Ethereum address. """
+    if isinstance(public_key, PublicKey):
+        public_key = public_key.format(compressed=False)
+    assert isinstance(public_key, bytes)
+    return sha3(public_key[1:])[-20:]
+
+
+def address_from_signature(data: bytes, signature: bytes, hasher: Hasher = sha3) -> Address:
+    """Convert an EC signature into an ethereum address"""
+    if not isinstance(signature, bytes) or len(signature) != 65:
+        raise InvalidSignature('Invalid signature, must be 65 bytes')
+    v = signature[-1]
+    # Support Ethereum's EC v value of 27,28 but also 0,1 to be in
+    # sync with the values accepted by the contracts:
+    # https://github.com/raiden-network/raiden-contracts/blob/aea36ce403605670edc23fe0d14cf422e2b8e69b/raiden_contracts/contracts/lib/ECVerify.sol#L27
+    if v in (27, 28):
+        signature = signature[:-1] + bytes([signature[-1] - 27])
+    elif v not in (0, 1):
+        raise InvalidSignature(f'Invalid signature. v value of {v} is illegal.')
+
+    try:
+        signer_pubkey = PublicKey.from_signature_and_message(signature, data, hasher=hasher)
+        return public_key_to_address(signer_pubkey)
+    except Exception as e:  # pylint: disable=broad-except
+        # coincurve raises bare exception on verify error
+        raise InvalidSignature('Invalid signature') from e
+
+
+def eth_recover(data: bytes, signature: bytes, hasher: Hasher = eth_sign_sha3) -> Address:
```

Comment:
nitpick: I think from our codestyle, there shouldn't be spaces around `=` here

Reply:
Hmm no. When it's a type function argument with a default value it needs spaces. Latest versions of `flake8` also complain for this:

`raiden/utils/signing.py:33:73: E252 missing whitespace around parameter equals`

Also check the pycodestyles [rule](https://pep8.readthedocs.io/en/latest/_modules/pycodestyle.html) for this.

```
Don't use spaces around the '=' sign in function arguments.

    Don't use spaces around the '=' sign when used to indicate a
    keyword argument or a default parameter value, except when
    using a type annotation.

    Okay: def complex(real, imag=0.0):
    Okay: return magic(r=real, i=imag)
    Okay: boolean(a == b)
    Okay: boolean(a != b)
    Okay: boolean(a <= b)
    Okay: boolean(a >= b)
    Okay: def foo(arg: int = 42):
    Okay: async def foo(arg: int = 42):

    E251: def complex(real, imag = 0.0):
    E251: return magic(r = real, i = imag)
    E252: def complex(real, image: float=0.0):
```